### PR TITLE
fix: avoid naked returns and unintended panics

### PR DESCRIPTION
**Pull request by swarmHaseeb**
### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] I have filled out the description and linked the related issues.

### Description

This PR removes naked returns from contract packages and adds a fix for unintended panics during tx Calls.
### Screenshots (if appropriate):

- - -
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3665)

### CODE REVIEW

**Review by notanatol — APPROVED** [review]

File: pkg/postage/postagecontract/contract.go
Comment (notanatol) [inline pkg/postage/postagecontract/contract.go:135]: I don't know if it makes sense, but since we expect exactly one result - we could just check if `len != 1` and specify in the error how many results we received?

**Comment by aloknerurkar** [comment]
So I am not sure if this change is needed. In the current testing of the release we have pretty much tested all this code time and again and never seen any panics. I would be categorized as nice to have, so I would hold this off before today's release.

**Review by vladopajic — APPROVED** [review]
LGTM

File: pkg/storageincentives/redistribution/redistribution.go
Comment (swarmHaseeb) [reply pkg/storageincentives/redistribution/redistribution.go:101]: that's good idea, thankyou

Comment (notanatol) [reply pkg/storageincentives/redistribution/redistribution.go:101]: maybe we could have a function `checkResults` that does both length check and type cast check...

Comment (vladopajic) [reply pkg/storageincentives/redistribution/redistribution.go:101]: one more note: 
it looks like `abi.ConvertType` method is used for similar usecase, but it can also panic if for same reason type from result does not match desired type

Comment (swarmHaseeb) [reply pkg/storageincentives/redistribution/redistribution.go:101]: thankyou, seems logical. Fixing those. 
@aloknerurkar we can merge this after releasing rc.

**Review by notanatol — APPROVED** [review]

File: pkg/postage/postagecontract/contract.go
Comment (mrekucci) [inline pkg/postage/postagecontract/contract.go:37]: `ErrUnexpectedResults` -> `ErrEmptyResult`; "unexpected empty results" -> "empty result".

File: pkg/storageincentives/redistribution/redistribution.go
Comment (mrekucci) [inline pkg/storageincentives/redistribution/redistribution.go:25]: The same as above.

File: pkg/storageincentives/staking/contract.go
Comment (mrekucci) [inline pkg/storageincentives/staking/contract.go:34]: The same as above.

File: pkg/util/abiutil/abiutil.go
Comment (mrekucci) [inline pkg/util/abiutil/abiutil.go:25]: Some documentation and tests would be great to have. The same for the function below.

Comment (aloknerurkar) [reply pkg/util/abiutil/abiutil.go:25]: I dont think these functions are being used anywhere?

File: pkg/util/abiutil/abiutil.go
Comment (vladopajic) [inline pkg/util/abiutil/abiutil.go:61]: could `ErrEmptyResults` be used here?

File: pkg/util/abiutil/abiutil.go
Comment (vladopajic) [inline pkg/util/abiutil/abiutil.go:75]: could `ErrEmptyResults` be used here?

File: pkg/util/abiutil/abiutil.go
Comment (vladopajic) [inline pkg/util/abiutil/abiutil.go:72]: maybe more suitable name for this function would be `UnpackBytes32`?

File: pkg/util/abiutil/abiutil.go
Comment (vladopajic) [inline pkg/util/abiutil/abiutil.go:44]: Naming convention for `UnpackSomething` had more sense while these functions were unpacking value from abi. Could functions with new approach have different naming; perhaps `ConvertSomething`?

File: pkg/util/abiutil/abiutil_test.go
Comment (vladopajic) [inline pkg/util/abiutil/abiutil_test.go:28]: This function name should be named `TestUnpack`, leaving out `BigInt` part since we are testing all `UnpackSomething` functions here.

Could all testes related to one method be in one subtest or at least close to each other?

File: pkg/util/abiutil/abiutil.go
Comment (istae) [inline pkg/util/abiutil/abiutil.go:51]: see Lint failure msg

File: pkg/util/abiutil/abiutil.go
Comment (istae) [inline pkg/util/abiutil/abiutil.go:32]: have we encountered any panic?

Comment (swarmHaseeb) [reply pkg/util/abiutil/abiutil.go:32]: if it's there, chances are we will

Comment (istae) [reply pkg/util/abiutil/abiutil.go:32]: I don't we need it, if we do proper error checking, we should be good.
We have not encountered a case of a panic, at least to my knowledge, and no operator has reported one either

File: pkg/storageincentives/staking/contract.go
Comment (istae) [inline pkg/storageincentives/staking/contract.go:236]: Why can't we use `abiutil` here ?

Comment (istae) [reply pkg/storageincentives/staking/contract.go:236]: this way, we don't need `ErrUnexpectedResults` in this pkg as well

**Review by istae — CHANGES_REQUESTED** [review]

File: pkg/util/abiutil/abiutil.go
Comment (istae) [reply pkg/util/abiutil/abiutil.go:33]: @swarmHaseeb ^^^

File: pkg/util/abiutil/abiutil.go
Comment (istae) [reply pkg/util/abiutil/abiutil.go:32]: @swarmHaseeb ^^^

File: pkg/util/abiutil/abiutil.go
Comment (swarmHaseeb) [reply pkg/util/abiutil/abiutil.go:33]: @istae well yes no one has reported any errors but if you look at test cases you can clearly see the edge cases that were originally missing, not today but tomorrow we can get an error. So this is proper error checking.

Comment (istae) [reply pkg/util/abiutil/abiutil.go:33]: so even the standard library packages have panics, and things panic when we write unrecoverable bugs and proper checking is not done

sticking a panic catcher here does not actually make the logic more "robust" it just sweeps potential bugs and problems with the code under the rug

**Review by istae — CHANGES_REQUESTED** [review]

File: pkg/util/abiutil/abiutil.go
Comment (vladopajic) [reply pkg/util/abiutil/abiutil.go:33]: `ConvertType` function will convert potential panic to error, therefore all panicking problems in code will not be disregarded.

Code which returns error instead of panic is generally more robust because developer have to handle error, while panic have to be "discover". Later comes up with a price if panicking code is discovered in production. 

---

If `ConvertType` is blocking this PR, we could come to common grounds, in order to move with this PR forward, and remove this this function and use one from `abi` package.

**Pull request closed without merging**
